### PR TITLE
Implement IgniteEntityEvent (Close #970)

### LINKED ISSUE #970: Implementation Request IgniteEntityEvent
1.10.2 versions dont have the IgniteEntityEvent fired when events burst into flame from fire, lava or sunlight (skeletons/zombies).  I dont know if this event was ever implimented in 1.8.9 and didn't make it.

DamageEntityEvent only goes so far as a workaround - it only registers living entities suffereing health damage ticks due to fire.

IgniteEntityEvent will capture armor stands being ignited by lava and fire  (as well as prevent need for listening to generic damage for living entities)

### PR DESCRIPTION
**Pull request by ImMorpheus**
I think I'm not getting the difference between "originalfireticks" and "fireticks".

https://github.com/SpongePowered/SpongeAPI/blob/bleeding/src/main/java/org/spongepowered/api/event/entity/IgniteEntityEvent.java

Is there a case when they are different ?

### CODE REVIEW

**Comment by parlough** [comment]
I'm assuming original means the ticks when the event was first thrown, so it won't change compared to the normal one which will when you set the ticks with the event. Event-gen-impl handles this.

**Comment by parlough** [comment]
Without looking to much into this, I'm assuming you need to still handle lightning as the source for the event and you probably need a check to make sure you're only using the cause stack manager if you can. Just return early if `((IMixinWorld) world).isFake()`

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java [hunk 1]
```diff
@@ -1360,4 +1363,26 @@ public boolean shouldTick() {
...
+                }
+                if (this.originalInFire != null) {
+                    frame.pushCause(this.originalInFire);
+                }
+                frame.pushCause(this.getLocation().getExtent());
+                IgniteEntityEvent event = SpongeEventFactory.
+                        createIgniteEntityEvent(frame.getCurrentCause(), this.fire, this.fire, this);
+                SpongeImpl.postEvent(event);
+
+                if (event.isCancelled()) {
```
Comment (parlough) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: `if (SpongeImpl.postEvent(event))`

**Comment by ImMorpheus** [comment]
Apparently, the event is not firing. 

During my tests I was injecting at HEAD, but I changed it to RETURN before commit because "this.fire" was not set at that point.

EDIT:

> I'm assuming you need to still handle lightning as the source for the event 

This is the cause when a skeleton is struck by a lightning:

```
Cause[Context=Context[], Stack={WorldServer{LevelName=world, DimensionId=0, DimensionType=minecraft:overworld}, EntityLightningBolt['unknown'/553, l='world', x=79.70, y=85.00, z=-11.30]}]
```

As of now there's no concept of "source", I tried 

```
PhaseData data = PhaseTracker.getInstance().getCurrentPhaseData();

data.context.getSource(Living.class).ifPresent(source ->
                        frame.addContext(EventContextKeys.IGNITER, source));
```

but this is only working for players (or other classes that extend Living), I can't add an EntityLightning as an IGNITER.

**Comment by ryantheleach** [comment]
The source would likely be the world or weather or something, unless it was triggered by API or Command.

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java [hunk 2]
```diff
@@ -1364,7 +1364,7 @@ public void setInvulnerable(boolean value) {
         this.invulnerable = value;
     }
 
-    @Inject(method = "setFire", at = @At(value = "RETURN"), cancellable = true)
+    @Inject(method = "setFire", at = @At(value = "HEAD"), cancellable = true)
```
Comment (ImMorpheus) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: Need help, setting this to RETURN doesn't fire the event at all.

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java [hunk 3]
```diff
@@ -1360,4 +1363,39 @@ public boolean shouldTick() {
     public void setInvulnerable(boolean value) {
         this.invulnerable = value;
     }
+
+    @Inject(method = "setFire", at = @At(value = "HEAD"), cancellable = true)
```
Comment (ImMorpheus) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: Need help, setting this to RETURN doesn't fire the event at all.

Comment (Cybermaxke) [reply src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: At RETURN is `isBurning` probably already returning `true`.

You will need to implement the injection differently if you want to support the overridden fire seconds. A `ModifyVariable` should do the trick. Inject into the seconds argument and return the modified seconds from the event. `isBurning` will also work at this point.

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java [hunk 4]
```diff
@@ -1360,4 +1363,39 @@ public boolean shouldTick() {
...
+                if (SpongeImpl.postEvent(event)) {
+                    ci.cancel();
+                }
+            }
+        }
+    }
+
+    private DamageSource originalLightning;
+
+    @Inject(method = "onStruckByLightning", at = @At(value = "HEAD"))
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: Where is this coming from?

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java (hunk 3 above)
Comment (gabizou) [reply src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: You can even get around by using a `@Redirect` with a field setter. This way you can cancel the return by sending the current fire seconds, or increase it based on whatever event is thrown.

**Review by gabizou — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java (hunk 4 above)
Comment (ImMorpheus) [reply src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: This was created to add the DamageSource when the entity is set on fire by a lightning. There's a field called `originalLava` (set inside `setOnFireFromLava()`) for lava and a field called `originalInFire` (set inside `dealFireDamage`) for fire, but nothing for lightning.

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java (hunk 3 above)
Comment (ImMorpheus) [reply src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: I was testing around and ended up with:
```
@Inject(method = "setFire",
            at = @At(value = "FIELD", target = "Lnet/minecraft/entity/Entity;fire:I", opcode = Opcodes.PUTFIELD),
            locals = LocalCapture.CAPTURE_FAILEXCEPTION,
            cancellable = true)
```
Everything is working, not sure what is better.

**Comment by parlough** [comment]
I feel like this event can fire quite a lot, especially with mobs during the day, and don't feel like it is something commonly listened to, so consider using ShouldFire for this, placing the check at the beginning of the `onFire` mixin.

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java (hunk 4 above)
Comment (gabizou) [reply src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: If it's coming from that, you need to be verifying where the modified lightning is being set. If it isn't being set, then this isn't doing anything special.

**Comment by ryantheleach** [comment]
One feature I was looking for a while back (That I don't particularly care for now) Was knowing when a fire was extinguished on an entity. Maybe it's more applicable to a data event. At least with the ignite event I can add them to a list of entities to poll the state of.

**Comment by Zidane** [comment]
@ImMorpheus 

Can you squash to one commit?

**Comment by ImMorpheus** [comment]
Done

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java [hunk 5]
```diff
@@ -1360,4 +1367,52 @@ public boolean shouldTick() {
     public void setInvulnerable(boolean value) {
         this.invulnerable = value;
     }
+
+    @Inject(method = "setFire",
+            at = @At(value = "FIELD", target = "Lnet/minecraft/entity/Entity;fire:I", opcode = Opcodes.PUTFIELD),
+            locals = LocalCapture.CAPTURE_FAILEXCEPTION,
+            cancellable = true)
+    public void onFire(int seconds, CallbackInfo ci, int i) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: public?

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java (hunk 4 above)
Comment (gabizou) [reply src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: Here's the thing, what I tried asking indirectly is that your'e not setting the `originalLightning` back to `DamageSource.LIGHTNING_BOLT`, nor are you making it a new `DamageSource`, so this injection is 100% useless without that logic existing to *reset* the damage sources.

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java [hunk 6]
```diff
@@ -1360,4 +1367,52 @@ public boolean shouldTick() {
...
+                    ci.cancel();
+                }
+            }
+        }
+    }
+
+    private DamageSource originalLightning;
+
+    @Inject(method = "onStruckByLightning", at = @At(value = "HEAD"))
+    public void onStruck(EntityLightningBolt lightningBolt, CallbackInfo ci) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: And this should be private.

**Review by gabizou — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java [hunk 7]
```diff
@@ -1360,4 +1368,63 @@ public boolean shouldTick() {
...
+
+                if (this.originalLava != null) {
+                    frame.addContext(EventContextKeys.DAMAGE_TYPE, DamageTypes.MAGMA);
+                    frame.pushCause(this.originalLava);
+                }
+                if (this.originalInFire != null) {
+                    frame.addContext(EventContextKeys.DAMAGE_TYPE, DamageTypes.FIRE);
+                    frame.pushCause(this.originalInFire);
+                }
+                if (this.originalLightning != null) {
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: And technically, lightning is considered `PROJECTILE`, not `FIRE`.

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java [hunk 8]
```diff
@@ -1360,4 +1368,63 @@ public boolean shouldTick() {
...
+            cancellable = true)
+    private void onFire(int seconds, CallbackInfo ci, int i) {
+        if (((IMixinWorld) world).isFake() || !ShouldFire.IGNITE_ENTITY_EVENT) {
+            return;
+        }
+        if (!this.isBurning()) {
+            try (CauseStackManager.StackFrame frame = Sponge.getCauseStackManager().pushCauseFrame()) {
+
+                if (this.originalLava != null) {
+                    frame.addContext(EventContextKeys.DAMAGE_TYPE, DamageTypes.MAGMA);
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: Lava blocks are fire, not magma (magma is specific to the magma block, not sure if we should change it).

**Review by gabizou — CHANGES_REQUESTED** [review]
Need to verify whether the damage types need to be changed in API or not for lightning. I consider lightning to be a projectile still, while lava blocks are fire, not magma damage types.

File: src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java [hunk 9]
```diff
@@ -1360,4 +1368,63 @@ public boolean shouldTick() {
...
+    private DamageSource originalLightning;
+
+    @Inject(method = "onStruckByLightning", at = @At(value = "FIELD", target = LIGHTNING_DAMAGESOURCE_FIELD, opcode = Opcodes.GETSTATIC))
+    private void onPreStruck(EntityLightningBolt lightningBolt, CallbackInfo ci) {
+        if (!this.world.isRemote) {
+            this.originalLightning = DamageSource.LIGHTNING_BOLT;
+            AxisAlignedBB bb = this.getEntityBoundingBox().grow(-0.001D, -0.001D, -0.001D);
+            Location<World> location = DamageEventHandler.findFirstMatchingBlock((net.minecraft.entity.Entity) (Object) this, bb, block ->
+                    block.getBlock() == Blocks.FIRE);
+            DamageSource.LIGHTNING_BOLT = new MinecraftBlockDamageSource("lightningBolt", location).setFireDamage();
```
Comment (gabizou) [inline src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntity.java]: Why a block and not the actual lightning bolt?

**Comment by gabizou** [comment]
Ok, so, I just went and made some changes on how all the DamageSource re-assignments work, they should all be redirects now (they were originally implemented before we had redirects available in mixin ~oh god... the horror~ added them).

To get an idea, review [this commit](https://github.com/SpongePowered/SpongeCommon/commit/07a334c4574ab942d0cdb453d7ed685f76d8c10b).

Update to that style and I'll merge the PR!

**Comment by ImMorpheus** [comment]
Oh, I see what you did. However take the `onSpongeRedirectForBlockDamageSource` method, the damagesource was assigned to this field https://github.com/SpongePowered/SpongeCommon/commit/07a334c4574ab942d0cdb453d7ed685f76d8c10b#diff-811f97d54d477fefea740d226d776a37L373 and I was using that field to establish the correct `EventContextKeys.DAMAGE_TYPE` and push the damagesource to the CauseStack https://github.com/SpongePowered/SpongeCommon/pull/1708/files#diff-811f97d54d477fefea740d226d776a37R1386.
Not sure how to do that now, am I missing something ?

**Comment by gabizou** [comment]
You can still do that now, just push to the cause stack when you re-assign/create the new `DamageSource`. Though, I'm understanding that the `attackEntity` method already pushes the `DamageSource` to the cause stack (can verify [here in the massive overwrite](https://github.com/SpongePowered/SpongeCommon/blob/cf5fa8b4e5d02a5e69b4bafa63a3fe2318033c8d/src/main/java/org/spongepowered/common/mixin/core/entity/MixinEntityLivingBase.java#L364-L549)). So, ideally you shouldn't have to push those sorts of things, but I'd double check the events with a test plugin maybe (I don't think we've written one for the test).

**Comment by Blayyke** [comment]
Any update on this? Would be incredibly useful to be able to use this event in my plugin

**Comment by ImMorpheus** [comment]
Sorry,  I've been busy with other things. 

Squashed and rebased, however this needs testing, there are a few oddities I've experienced.

**Comment by ImMorpheus** [comment]
Should be good. 
Not sure what to do with creative players: if you switch to creative mode and start swimming in a lava lake, the server spams the event (a lot). That's because it's calling `Entity#extinguish()` if you're in creative before setting you on fire again. This is not happening with invulnerable entity.

**Comment by gabizou** [comment]
Had to make some changes to get the values back from plugins and setting them to the entity again. If you want to look at the commits shortly after to see how it actually works (with a very simple test plugin).

**Pull request merged**